### ISSUE TITLE
awaitSchemaAgreement not ignoring ghost node

### ISSUE THREAD

**Issue by fesiqueira** [gocql/gocql#1736]
Please answer these questions before submitting your issue. Thanks!

### What version of Cassandra are you using?
3.11.11

### What version of Gocql are you using?
v1.6.0

### What version of Go are you using?
1.21.6

### What did you do?
Running migrations that use `WaitForSchemaAgreement`

### What did you expect to see?
The migrations to finish successfully

### What did you see instead?
```
gocql: cluster schema versions not consistent: [<uuid1> <uuid2>]
```

---

If you are having connectivity related issues please share the following additional information

### Describe your Cassandra cluster
I have a cluster of 50+ nodes. One of them was replaced a few days ago but the `system.peers` was left in an inconsistent state, where the replaced node is still showing in there with a different `schema_version`.

The `nodetool status` command prints out the "correct" output, where the old node is NOT shown.

The `system.peers` table shows the old node and all columns are filled with some value, like:
```
peer         | rpc_address  | schema_version
10.100.XX.XX | 10.100.XX.XX | <uuid2> 
```

The old node is unreachable.

**Comment by fesiqueira** [comment]
Hi @joao-r-reis,

I wanted to check in on this PR/issue since it has been open for a while. Does it make sense to keep it open? If so, could you review it when you have a chance? If it's not relevant, I’m happy to close it.

Thanks for your time!

**Comment by joao-r-reis** [comment]
Yeah it makes sense to keep it open, right now we're prioritizing some other issues for the upcoming 2.0 release but this is definitely something that should be fixed at some point. Can you create a [JIRA](https://issues.apache.org/jira/projects/CASSGO/issues)?

**Comment by fesiqueira** [comment]
Done! Opened [CASSGO-52](https://issues.apache.org/jira/browse/CASSGO-52).

**Comment by OleksiienkoMykyta** [comment]
@joao-r-reis, so we just leave it to the v2.x?

**Comment by fesiqueira** [comment]
Hey! I just wanted to check in, as I ran into this issue again this week.

I noticed the v2.x release seems to be approaching (with a release candidate out), so I was wondering if there's been any update on this issue, or any feedback on the PR I opened for it. I believe it might not be the optimal fix, but I'd appreciate your thoughts on it.

Thanks a lot!

**Comment by dkropachev** [comment]
> Hey! I just wanted to check in, as I ran into this issue again this week.
> 
> I noticed the v2.x release seems to be approaching (with a release candidate out), so I was wondering if there's been any update on this issue, or any feedback on the PR I opened for it. I believe it might not be the optimal fix, but I'd appreciate your thoughts on it.
> 
> Thanks a lot!

Could you please provide full record from system.peers and/or system.peers_v2 ?

**Comment by fesiqueira** [comment]
> Could you please provide full record from system.peers and/or system.peers_v2 ?

@dkropachev I can't pull it right now, but the `system.peers` looks something like:

```
peer         | rpc_address  | schema_version
10.100.XX.01 | 10.100.XX.01 | <uuid1> 
10.100.XX.02 | 10.100.XX.02 | <uuid1> 
10.100.XX.03 | 10.100.XX.03 | <uuid1> 
10.100.XX.04 | 10.100.XX.04 | <uuid2> <-- Unreachable address/node, which was replaced by another node.
10.100.XX.05 | 10.100.XX.05 | <uuid1> <-- Brand-new node, replacement of the address/node above.
```

Is this helpful?

**Comment by dkropachev** [comment]
I think everyone know how roughly it looks like, in order to not reproduce this case it would be handy to have exact record to see if driver can guess from it if node is down or off the cluster.

**Comment by joao-r-reis** [comment]
I don't think the 2.0 release will have anything to make this better but I'm not sure what you're looking for here. If there is a ghost node on system.peers then the driver will always consider it a valid node so the awaitschemaagreement function will also use it. The fix would be to resolve the issue with the system.peers table but I can't help much with that, maybe asking C* folks in ASF slack or the C* mailing list will help you.

You can try filtering out a host by providing a HostFilter as a workaround to see if that works too

**Comment by joao-r-reis** [comment]
Oh maybe you're looking for something like [this](https://github.com/apache/cassandra-java-driver/blob/e2c7ad4d11555eeacce6bd436547b403f83eb24f/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/SchemaAgreementChecker.java#L192-L211). If gocql doesn't ignore DOWN nodes (along with other checks in that java driver method) then that is something we can add for sure

**Comment by dkropachev** [comment]
> > Could you please provide full record from system.peers and/or system.peers_v2 ?
> 
> [@dkropachev](https://github.com/dkropachev) I can't pull it right now, but the `system.peers` looks something like:
> 
> ```
> peer         | rpc_address  | schema_version
> 10.100.XX.01 | 10.100.XX.01 | <uuid1> 
> 10.100.XX.02 | 10.100.XX.02 | <uuid1> 
> 10.100.XX.03 | 10.100.XX.03 | <uuid1> 
> 10.100.XX.04 | 10.100.XX.04 | <uuid2> <-- Unreachable address/node, which was replaced by another node.
> 10.100.XX.05 | 10.100.XX.05 | <uuid1> <-- Brand-new node, replacement of the address/node above.
> ```
> 
> Is this helpful?

BTW, did you try `nodetool assassinate <ip>` ? and/or `DELETE FROM system.peers WHERE peer = '<ip>';`

**Comment by fesiqueira** [comment]
> Oh maybe you're looking for something like [this](https://github.com/apache/cassandra-java-driver/blob/e2c7ad4d11555eeacce6bd436547b403f83eb24f/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/SchemaAgreementChecker.java#L192-L211). If gocql doesn't ignore DOWN nodes (along with other checks in that java driver method) then that is something we can add for sure

@joao-r-reis I believe that's what I need (similar to what I added in #1738), and if it makes sense for gocql to have these checks I'd like to address any feedback on that PR and keep working on it.

The node gocql keeps complaining about the schema version is down and it will never be up (or reachable) again as it was destroyed and replaced by a new instance.

@dkropachev I tried both  `nodetool removenode` and `nodetool assassinate`.

While `assassinate` removed the bad row from `system.peers` I would still like gocql to ignore it, as in the near future we will make some changes to our infrastructure and login in to the cluster would become more restricted.

**Comment by dkropachev** [comment]
> [@dkropachev](https://github.com/dkropachev) I tried both `nodetool removenode` and `nodetool assassinate`.
> 
> While `assassinate` removed the bad row from `system.peers` I would still like gocql to ignore it, as in the near future we will make some changes to our infrastructure and login in to the cluster would become more restricted.

The question is how driver can distinguish ghost nodes from normal, me, looking at the schema of `system.peers` it is not clear, unless you provide exact record from that table for ghost node it would be hard to tell how to do that or is it even possible.

**Comment by fesiqueira** [comment]
> The question is how driver can distinguish ghost nodes from normal, me, looking at the schema of `system.peers` it is not clear, unless you provide exact record from that table for ghost node it would be hard to tell how to do that or is it even possible.

I think I created some confusion by calling it a "ghost node", but just by looking at the `system.peers` table is not really possible to tell which record is not supposed to be there for sure. 

It has nothing particularly distinctive from other records, with the exception that I know it should not be there because I manually destroyed the Cassandra node with said IP.

What I thought (and maybe it's the only way to implement it, not sure) is something like João [linked](https://github.com/apache/cassandra-java-driver/blob/e2c7ad4d11555eeacce6bd436547b403f83eb24f/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/SchemaAgreementChecker.java#L199), where `gocql` would check if the node is up, and if not its schema version would not be checked against the others. What do you think?

**Comment by joao-r-reis** [comment]
Yeah I'm going to rename this issue to "not ignoring DOWN/invalid node", whether it's a ghost node or not doesn't really matter because the driver can't tell

**Comment by eriksw** [comment]
Please note that it is possible to filter bogus `system.peers` entries as long as you have access to up/down status of nodes.

For a bogus entry left behind after replacing a node, for each of its tokens, there will be some other node that owns the token and is up.

If all distinct known tokens are accounted for by hosts that are up and agree, `awaitSchemaAgreement` should return nil.


```
 peer          | data_center | host_id                              | preferred_ip | rack | release_version | rpc_address   | schema_version                       | tokens
---------------+-------------+--------------------------------------+--------------+------+-----------------+---------------+--------------------------------------+--------------------------------------------------------------------------------------------------
...
  10.100.12.56 |    fcevent1 | c048b2bf-3c6a-47e2-a386-2f5d3966a3bf |         null |    b |         3.11.11 |  10.100.12.56 | 269a255f-a4ad-312f-9ac8-6ddd1d7396b4 |   {'-4136344956821905473', '-7691186262693016981', '4222335951577735102', '7296793297195993708'}
  10.100.12.67 |    fcevent1 | d0918a84-f012-4267-b0e7-09d100569a61 |         null |    a |         3.11.11 |  10.100.12.67 | 269a255f-a4ad-312f-9ac8-6ddd1d7396b4 |     {'2300800110566323476', '5759564624386864404', '8065407633600558358', '9218329138207405334'}
  10.100.12.41 |    fcevent1 | 153cf2ce-4ace-44ab-95b1-fe7343a5aa14 |         null |    b |         3.11.11 |  10.100.12.41 | 1fb847cb-875c-3aec-ae37-b7a68fab1896 |    {'-3655960996569052565', '-8267647014996440468', '5567411040285723241', '955725021858335338'}
 10.100.12.194 |    fcevent1 | 1fdceb8e-8cd6-465a-a2d3-26dac5e3fd6d |         null |    b |         3.11.11 | 10.100.12.194 | 269a255f-a4ad-312f-9ac8-6ddd1d7396b4 |    {'-3655960996569052565', '-8267647014996440468', '5567411040285723241', '955725021858335338'}
  10.100.12.30 |    fcevent1 | 05aa4f26-12eb-4d5d-b9db-88f9d800f923 |         null |    c |         3.11.11 |  10.100.12.30 | 269a255f-a4ad-312f-9ac8-6ddd1d7396b4 |   {'-2695193076063346753', '-3848114580670193728', '4894873495931729171', '5663487832336293822'}
 10.100.12.184 |    fcevent1 | 6eb54f63-f1eb-426f-8d92-515b1e323945 |         null |    b |         3.11.11 | 10.100.12.184 | 269a255f-a4ad-312f-9ac8-6ddd1d7396b4 |    {'-2503039491962205590', '-773657235051935126', '4606643119780017427', '8834021970005123009'}
...
```
